### ISSUE TITLE
Refreshing then hitting the back button causes 404

### ISSUE THREAD

**Issue by hossameldeen** [gatsbyjs/gatsby#29469]
## Description

If you refresh a page, then click on the back button, you get 404 instead of getting to the page you were on.

#### Links

- Reproduction on Codesandbox: https://codesandbox.io/s/github/hossameldeen/Gatsby-site
- Source code link: https://github.com/hossameldeen/Gatsby-site

#### Steps
- Clone the repo, run `gatsby develop`, and open `http://localhost:8000` (or just open codesandbox link above)
- You should see a link `click to go to /another-page`
![image](https://user-images.githubusercontent.com/6719916/107723566-24c57700-6cea-11eb-9fd1-f11efe75bcfb.png)
- Click on the link.
- You should now see a page with some text.
![image](https://user-images.githubusercontent.com/6719916/107723601-3b6bce00-6cea-11eb-9375-38615e77c4e6.png)
- Click on refresh button in the browser.
- Click on the back button in the browser.

### Expected result

Go back to the homepage `http://localhost:8000` and show its contents (the link)

### Actual result

Getting 404
![image](https://user-images.githubusercontent.com/6719916/107723678-7cfc7900-6cea-11eb-80aa-2dfa7fa5d47f.png)

### Environment

```
  System:
    OS: Windows 10 10.0.19041
    CPU: (12) x64 Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
  Binaries:
    Node: 12.19.0 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.22.5 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.14.8 - C:\Program Files\nodejs\npm.CMD
  Browsers:
    Edge: Spartan (44.19041.423.0), Chromium (88.0.705.63)
  npmPackages:
    gatsby: ^2.32.3 => 2.32.3 
```

### More info

- Seems to be the same issue as #7261
- Credits of finding the issue to @mostafa-zaied

**Comment by vladar** [comment]
I think this only happens in development? Have you tried it with `gatsby build` / `gatsby serve`?

**Comment by hossameldeen** [comment]
@vladar You're right, it doesn't happen when I use `gastby build && gatsby serve`. Is this an intended behavior in development mode or just a bug?

**Comment by hossameldeen** [comment]
Yes, I want to keep this issue open as it's probably a bug. Feel free to close this issue if this low-priority due to it being in development mode only.

**Comment by cjones26** [comment]
> Yes, I want to keep this issue open as it's probably a bug. Feel free to close this issue if this low-priority due to it being in development mode only.

I am also seeing the same behavior.  It definitely appears to be a bug.  Out of curiosity, are you using `index.js` or Typescript?  I'm using `index.tsx` I'm wondering if that may have anything to do with it.

**Comment by Eytyy** [comment]
> > Yes, I want to keep this issue open as it's probably a bug. Feel free to close this issue if this low-priority due to it being in development mode only.
> 
> I am also seeing the same behavior. It definitely appears to be a bug. Out of curiosity, are you using `index.js` or Typescript? I'm using `index.tsx` I'm wondering if that may have anything to do with it.

I'm not using Typescript and having the same issue.

**Comment by hossameldeen** [comment]
@cjones26 I don't remember exactly since it's been a while, but as far as I remember:

- Our project is in TypeScript, so it happens with TypeScript.
- The [reproduction repo](https://github.com/hossameldeen/Gatsby-site) is in JavaScript, so I guess it happens with Javascript as well :(

**Comment by VasilyMur** [comment]
I see the same issue with official Gatsby starter.
Once I hit refresh button and hit either back or forward button in the browser - I get 404.
At the same time the 404 page lists the "unavailable" page under Pages.

**Comment by skyrocketeer** [comment]
I am having the same issue
However if you refresh the page after hitting the back browser back button, it works as expected.

Any solution or workaround here. It seems like a bug

**Comment by hossameldeen** [comment]
@skyrocketeer In case you didn't notice, we've reached the conclusion that this bug happens in development mode only. But if you use something like `gatsby build` or `gatsby build && gatsby serve`, the bug shouldn't happen. Hope that helps :)

**Comment by jdahdah** [comment]
Also getting 404 pages in `develop` when using `navigate()` (such as after a form submission).

**Comment by vinaysharma14** [comment]
@hossameldeen did you find any solution for this?

**Comment by hossameldeen** [comment]
@vinaysharma14 Nope :( ... And in case you didn't notice: This happens only in the development mode, it seems. So, for the production build with `gatsby build`/`gatsby build && gatsby serve`, this problem shouldn't occur. Hope that reduces the issue's impact for you :)

**Comment by vinaysharma14** [comment]
@hossameldeen Yes, I did notice that it happens only in development mode. But it's quite weird as I'm not at all able to reproduce it today.

**Comment by vinaysharma14** [comment]
@hossameldeen Yes, I did notice that it happens only in development mode. But it's quite weird as I'm **unable to reproduce** it today at all.

**Comment by rodikh** [comment]
I manage to reproduce this issue not only by hitting back after a refresh, but also when clicking on some <Link> links quickly after the page loads. 
Makes me think these two issues are related, and that they are both related to <Link> prefetching.

This happens with various gatsby versions including `3.9.1`.

**Comment by dawidjaniga** [comment]
Hey,
I encountered the same issue with Gatsby 3.11.1
This is the same problem as https://github.com/gatsbyjs/gatsby/issues/7261

**Comment by mickeyryan42** [comment]
Has this issue been investigated?  It still seems to be present in 4.13.0.

It's true that the impact is lower because it only causes a 404 in development, but I believe it is still relevant as it causes a good bit of frustration in development, especially in trying to develop features that involve route changes.

**Comment by MartijnHols** [comment]
I ran into this in 5.5.0. I did some quick investigation where I found the page is being rendered partially as expected, and then a frame (or at most a fraction of a second) later, it's replaced by the dev-404 page. Unfortunately I don't have more time available to investigate this bug.

**Comment by aprilmintacpineda** [comment]
I get this error in dev mode with v5.8.1.

**Comment by kndshein** [comment]
Still getting this error w/ v5.13.
